Ex1/week6_2.py

```diff
@@ -0,0 +1,161 @@
+    return num1 + num2
+
+
+def subtract(num1, num2):
+    return num1 - num2
+
+
+def multiply(num1, num2):
+    return num1 * num2
+
+
+def divide(num1, num2):
+    return num1 / num2
+
+
+def calc(num1, num2, math_sign):
+    """
+    Gets 2 numbers and a math sign (+, -, * or /) and return the result by using a dictionary of functions
+    :param num1: First number
+    :param num2: Second number
+    :param math_sign: A math sign
+    :return:
+    """
+    dictionary = {
+        '+': add(num1, num2),
+        '-': subtract(num1, num2),
+        '*': multiply(num1, num2),
+        '/': divide(num1, num2)
+    }
+    return dictionary[math_sign]
+
+
+# From here to here
+
+
+def apply(func, iterable):
+    """
+    A generator that gets a function and an iterable and on each item of the iterable return the result
+    of the item after activate the function on it
+    :param func: A function
+    :param iterable: An iterable
+    :return: Activation of the function on each item of the iterable
+    """
+    for item in iterable:
+        yield func(item)
+
+
+# map
+
+
+def generate_first_names(lst_full_names):
+    """
+    Generator function that get a list of full names and returns an iterator that point only to the first names
+    :param lst_full_names: A list of full names (using map)
+    :return: An iterator of the first names only
+    """
+    return map(lambda name:  name.split(' ', 1)[0], lst_full_names)
+
+
+# filter
+
+
+def generate_only_polyndromes(lst_of_words):
+    """
+    Generator that get list of words and returns a generator of the words that are polyndromes only
+    :param lst_of_words: A list of words (using filter)
+    :return: An iterator on the words that are polyndromes
+    """
+    return filter(lambda word: word == word[::-1], lst_of_words)
+
+
+# Personally adjusted filter
+
+
+def my_filter(func, iterable):
+    """
+    A generator function that implement the known function "filter of python
+    :param func: A function
+    :param iterable: An iterable
+    :return: An iterator of the items in the iterable that return True from the function sent
+    """
+    for item in iterable:
+        if func(item):
+            yield item
+
+
+# Stay ? positive
+
+
+def get_positive_numbers(inp):
+    """
+    A function that gets an input of numbers seperated by ',' and return a list of the positive numbers
+    :param inp: The input from the user
+    :return: list of all positives
+    """
+    return list(filter(lambda x: int(x) > 0, list(map(int, inp.split(",")))))
+
+
+# 2000 running
+
+
+def timer(f, *args):
+    """
+    A function that check out the time that to the function passed to run on the arguments passed
+    :param f: A function
+    :param args: A list of arguments
+    :return: The time that the function passed run on the arguments passed
+    """
+    start_time = time.time()
+    f(args)
+    return time.time() - start_time
+
+
+if __name__ == '__main__':
+    print(calc(5, 2, '+'))
```

Comment:
I would split all this logic into separate functions as well, this is a bit messy